### PR TITLE
[Approach #1] Improve IsDiagnosticAnalyzerSuppressed performance

### PR DESCRIPTION
**Pull request by mavasani**
Fixes [AB#1799782](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1799782). This fixes a long outstanding performance issue for CodeCleanup as well as command line builds.

`AnalyzerConfigSet` now exposes a set of all unique option keys and configured diagnostic IDs, which are in turn exposed by `SyntaxTreeOptionsProvider` and `AnalyzerConfigOptionsProvider`. These keys enable us to do a fast check in `IsDiagnosticAnalyzerSuppressed` to see if a diagnostic ID is never configured in an editorconfig/globalconfig in the entire `AnalyzerConfigSet` for the compilation. The slow path walks all the config files for each tree to determine if there is any entry which enables/unsuppresses the diagnostic ID.

### CODE REVIEW

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Holds keys for all options in an <see cref="AnalyzerConfigSet"/>.
+    /// </summary>
+    public readonly struct AnalyzerConfigOptionKeys
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs:16]: going to bring to API review i presume?

Comment (mavasani) [reply src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs:16]: Yes, this is just for experimental purpose. I want to first verify all tests pass (hence the draft PR), then measure perf and then bring it for API review. Finally, mark the PR ready for review. Feel free to review the changes now if you'd like though.

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        internal AnalyzerConfigOptionKeys(ImmutableHashSet<string> configuredDiagnosticIds, ImmutableHashSet<string> analyzerOptionKeys)
+        {
+            ConfiguredDiagnosticIds = UpdateComparerIfNeeded(configuredDiagnosticIds);
+            AnalyzerOptionKeys = UpdateComparerIfNeeded(analyzerOptionKeys);
+        }
+
+        private static ImmutableHashSet<string> UpdateComparerIfNeeded(ImmutableHashSet<string> keys)
+            => keys.KeyComparer == AnalyzerConfig.Section.PropertiesKeyComparer
+                ? keys
+                : keys.WithComparer(AnalyzerConfig.Section.PropertiesKeyComparer);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs:38]: is this called?

File: src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        internal AnalyzerConfigOptionKeys WithAdditionalAnalyzerConfigOptionKeys(IEnumerable<string> optionKeys)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/AnalyzerConfigOptionKeys.cs:40]: can this argument be mor strongly typed?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs [hunk 4]
```diff
@@ -400,6 +408,37 @@ internal bool IsSupportedDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic diag
 
                 return false;
             }
+
+            static bool isNeverConfiguredInEditorConfig(
+                DiagnosticDescriptor descriptor,
+                SyntaxTreeOptionsProvider treeOptionsProvider,
+                AnalyzerConfigOptionsProvider configOptionsProvider,
+                CancellationToken cancellationToken)
+            {
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs:417]: Method needs comments

**Comment by CyrusNajmabadi** [comment]
oh sorry.  i missed this is in draft.  will review when you want in the future.

**Comment by sharwell** [comment]
This is one of two identified approaches for this. There's a possibility that a second approach (exposing a helper on CompilationWithAnalyzers that filters a colleciton of analyzers down to the set of unsuppressed ones) would not require API changes within the analyzer config features. The investigation is ongoing.

**Comment by mavasani** [comment]
> This is one of two identified approaches for this. There's a possibility that a second approach (exposing a helper on CompilationWithAnalyzers that filters a colleciton of analyzers down to the set of unsuppressed ones) would not require API changes within the analyzer config features. The investigation is ongoing.

Hmm, adding an overload for the below API (which is soon going to be deprecated with #67671) which takes an `AnalyzerConfigSet` as an additional input can likely avoid other public API changes and threading around of the option keys. I'll investigate that approach and open another draft PR, and then create a tracking issue for API review to discuss which API shape feels better. However, I am first going to measure perf to confirm we are indeed getting good performance improvements here to invest in these efforts.

https://github.com/dotnet/roslyn/blob/d2dc53dbb0c6c67f206a74aa4d323d11cb2e107d/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L1061-L1073

**Comment by mavasani** [comment]
> This is one of two identified approaches for this. There's a possibility that a second approach (exposing a helper on CompilationWithAnalyzers that filters a colleciton of analyzers down to the set of unsuppressed ones) would not require API changes within the analyzer config features. The investigation is ongoing.

@sharwell I was able to come up with a much smaller public API change + implementation changes to get the same functionality. Opened #67974

**Comment by mavasani** [comment]
Performance measurements for code cleanup shows that this has been addressed with #67591. I am also unable to see any performance improvements for command line builds with this change. So, I am going to close this PR until we again see performance feedback where this method shows up in traces. [AB#1799782](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1799782) has been marked as fixed pending release.

**Pull request closed without merging**